File: src/main/java/com/educative/grokking/exercises/SearchinRotatedSortedArray.java
package com.educative.grokking.exercises;

import java.util.List;

/*
Given a sorted integer array, nums, and an integer value, target, the array is rotated by some arbitrary number.
Search and return the index of target in this array. If the target does not exist, return -1.
 */
public class SearchinRotatedSortedArray {
    public static int binarySearchRotated(List<Integer> nums, int target) {

        var left = 0;
        var right = nums.size() - 1;
        while (left <= right) {
            var mid = left + (right - left) / 2;
            if (nums.get(mid) == target) return mid;

            if (nums.get(left) <= nums.get(mid)) {
                if (nums.get(left) <= target && target <= nums.get(mid)) {
                    right = mid - 1;
                } else {
                    left = mid + 1;
                }
            } else {
                if (nums.get(mid) <= target && target <= nums.get(right)) {
                    left = mid + 1;
                } else {
                    right = mid - 1;
                }
            }
        }
        return -1;
    }
}

File: src/main/kotlin/neetcode/SearchinRotatedSortedArray.kt
package neetcode

// 33. Search in Rotated Sorted Array
// https://leetcode.com/problems/search-in-rotated-sorted-array/description/
class SearchinRotatedSortedArray {
    fun search(nums: IntArray, target: Int): Int {
        if (nums.size == 1) {
            return if (nums[0] == target) 0 else -1
        }

        var l = 0
        var r = nums.size - 1
        while (l <= r) {
            val m = l + (r - l) / 2

            if (nums[m] == target) {
                return m
            }
            if (nums[m] >= nums[l]) {
                // we are in the left sorted part, no pivot here, pivot is after m
                if (target <= nums[m] && target >= nums[l]) {
                    r = m - 1
                } else {
                    l = m + 1
                }
            } else {
                // we are in the right sorted part, pivot is here, pivot is before m
                if (target > nums[m] && target <= nums[r]) {
                    l = m + 1
                } else {
                    r = m - 1
                }
            }
        }

        return -1;
    }
}